convex/lp/simplex.go

```diff
@@ -0,0 +1,608 @@
+	// at an initial constraint vertex and successively moving to adjacent constraint
+	// vertices. At every vertex, the set of non-zero x values is the "basic
+	// feasible solution". The list of non-zero x's are maintained in basicIdxs,
+	// the respective columns of A are in ab, and the actual non-zero values of
+	// x are in xb.
+	//
+	// The LP is equality constrained such that A * x = b. This can be expanded
+	// to
+	//  ab * xb + an * xn = b
+	// where ab are the columns of a in the basic set, and an are all of the
+	// other columns. Since each element of xn is zero by definition, this means
+	// that for all feasible solutions xb = ab^-1 * b.
+	//
+	// Before the simplex algorithm can start, an initial feasible solution must
+	// be found. If initialBasic is non-nil a feasible solution has been supplied.
+	// Otherwise the "Phase I" problem must be solved to find an initial feasible
+	// solution.
+
+	var basicIdxs []int // The indices of the non-zero x values.
+	var ab *mat64.Dense // The subset of columns of A listed in basicIdxs.
+	var xb []float64    // The non-zero elements of x. xb = ab^-1 b
+
+	if initialBasic != nil {
+		// InitialBasic supplied. Panic if incorrect length or infeasible.
+		if len(initialBasic) != m {
+			panic("lp: incorrect number of initial vectors")
+		}
+		ab = extractColumns(A, initialBasic)
+		xb, err = initializeFromBasic(ab, b)
+		if err != nil {
+			panic(err)
+		}
+		basicIdxs = make([]int, len(initialBasic))
+		copy(basicIdxs, initialBasic)
+	} else {
+		// No inital basis supplied. Solve the PhaseI problem.
+		basicIdxs, ab, xb, err = findInitialBasic(A, b)
+		if err != nil {
+			return math.NaN(), nil, nil, err
+		}
+	}
+
+	// basicIdxs contains the indexes for an initial feasible solution,
+	// ab contains the extracted columns of A, and xb contains the feasible
+	// solution. All x not in the basic set are 0 by construction.
+
+	// nonBasicIdx is the set of nonbasic variables.
+	nonBasicIdx := make([]int, 0, n-m)
+	inBasic := make(map[int]struct{})
+	for _, v := range basicIdxs {
+		inBasic[v] = struct{}{}
+	}
+	for i := 0; i < n; i++ {
+		_, ok := inBasic[i]
+		if !ok {
+			nonBasicIdx = append(nonBasicIdx, i)
+		}
+	}
+
+	// cb is the subset of c for the basic variables. an and cn
+	// are the equivalents to ab and cb but for the nonbasic variables.
+	cb := make([]float64, len(basicIdxs))
+	for i, idx := range basicIdxs {
+		cb[i] = c[idx]
+	}
+	cn := make([]float64, len(nonBasicIdx))
+	for i, idx := range nonBasicIdx {
+		cn[i] = c[idx]
+	}
+	an := extractColumns(A, nonBasicIdx)
+
+	bVec := mat64.NewVector(len(b), b)
+	cbVec := mat64.NewVector(len(cb), cb)
+
+	// Temporary data needed each iteration. (Described later)
+	r := make([]float64, n-m)
+	move := make([]float64, m)
+
+	// Solve the linear program starting from the initial feasible set. This is
+	// the "Phase 2" problem.
+	//
+	// Algorithm:
+	// 1) Compute the "reduced costs" for the non-basic variables. The reduced
```

Comment:
I would prefer to have these comments distributed to corresponding places inside the loop body. The body is not so big and having the descriptions closer to the actual code is clearer, at least for me. Up to you.